strateos/transcriptic
# feature: add filtering of instructions on the exec command

## Summary:

 - Add a `--exclude/-e` and `--include/-i` flags on the exec command to filter instructions. The filters are being validated then forwarded to SCLE where the filtering is done: https://github.com/strateos/lab/pull/1527
 - Minor: use email address to specify who sent the payload.

Require: https://github.com/strateos/lab/pull/1526 to be merged in lab as well
### Usage:
```
cat autoprotocol.json | \
transcriptic exec -a lab.core.strateos.com/haven/wctest -w wc4-mcx1 -e 1-9 -i 7 -e op:provision
```
This command will first filter out instruction from index 1 to 9 (`-e 1-9`) and the provisions (`-e op:provision`). Then add back the instruction at index 7 (`-i 7`).

The order of `-i/-e` don't matter, it is first removing all the `-e`, then adding back all the `-i`.

### Filters

Possible filter:
 - single index: 123
 - range: 123-456
 - key:value in the instruction: `op:provision`, or `x_human!:false`. The ! means that if the key (x_human) is not in the instruction, then the filter matches. Otherwise, without the '!' a missing key means that there is no match.

Filter example:
  - { "op": "seal", "x_human": false } and { "op": "seal" } match `x_human!:false`
  - { "op": "seal" } does not match`x_human:true`, nor `x_human:false`

### Test plan
Add unit tests

#### Manual testing

```
internal_protocols/clients/sd2/current_protocols/growth_curve [master*]$ transcriptic preview GrowthCurve | \
transcriptic exec -a lab.core.strateos.com/haven/wctest -w wc4-mcx1 -e x_human:true
Sending request to http://13.52.223.111:9000/testRun
Success. View http://lab.core.strateos.com/haven/wctest/dashboard to see the scheduling outcome.
```
or
```
internal_protocols/clients/sd2/current_protocols/growth_curve [master*]$ transcriptic preview GrowthCurve | \
transcriptic exec -a lab.core.strateos.com/haven/wctest -w wc4-mcx1 -e op:provision
Sending request to http://13.52.223.111:9000/testRun
Success. View http://lab.core.strateos.com/haven/wctest/dashboard to see the scheduling outcome.
```

![image](https://user-images.githubusercontent.com/62562051/122622012-2c0be100-d04c-11eb-8ef8-b0c2263d15ac.png)

Bonus with the newest features not yet merged:
  - UI upgrade:  https://github.com/strateos/lab-frontend/pull/46
  - Rescheduling with new starting times: https://github.com/strateos/lab/pull/1521
  - Extract device set from file added to the docker container: https://github.com/strateos/lab/pull/1526

![image](https://user-images.githubusercontent.com/62562051/122627945-d1807e00-d067-11eb-963f-b73f2a02be6a.png)
